Add support for Bottlerocket nodgroups with NVIDIA accelerated hardware

arnaldo2792 opened the pull request:
### Description
Closes #4760

This pull request adds support for Bottlerocket nodegroups with NVIDIA accelerated hardware, the changes are as follows:

```
    Differentiate among accelerated hardware providers

    `IsGPUInstanceType` groups all together NVIDIA and inferentia instances
    which makes it difficult to reuse for OS'es that don't vend one image
    with support for multiple accelerated hardware.

    With this change, the instances with support for NVIDIA GPUs are
    separated in its own function (like inferentia instances), which is
    called by `IsGPUInstanceType`.
```

```
    Autoresolve SSM parameters for Bottlerocket GPU nodegroups

    This ensures that the correct SSM parameter is used to fetch the AMI id
    for Bottlerocket nodegroups with NVIDIA hardware.
```

```
    Skip NVIDIA device plugin daemonset in Bottlerocket node groups

    The Bottlerocket NVIDIA AMI already provides the NVIDIA device plugin,
    so Bottlerocket node groups don't require the NVIDIA device plugin
    daemonset.

    This change makes sure that clusters with only Bottlerocket nodegroups
    don't get the NVIDIA device plugin daemonset. If any of the nodegroups
    in the cluster's configuration uses an OS that requires the device
    plugin, it will be installed.
```

### Manual testing

- With the following configuration file, I verified that the correct AMI was auto resolved based on the `amiFamily` and the `instanceType`, and that the device plugin daemonset wasn't installed:

```yaml
---
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: bottlerocket-1-21
  region: us-west-2
  version: '1.21'

nodeGroups:
  - name: al2-without-plugin-ng
    instanceType: m5.large
    desiredCapacity: 1
  - name: br-without-plugin
    instanceType: p3.2xlarge
    desiredCapacity: 1
    amiFamily: Bottlerocket
    iam:
       attachPolicyARNs:
          - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
          - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
          - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
          - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
    ssh:
        allow: true
        publicKeyName: <>
    bottlerocket:
      settings:
        motd: "Hello from eksctl!"
        host-containers:
          admin:
            enabled: true
```

```sh
~ ❯ eksctl get nodegroup --cluster bottlerocket-1-21 -o json | jq '.[].Name, .[].Type, .[].ImageID' -r
br-without-plugin
unmanaged
ami-06e0114f5eb33d00c
~ ❯ kubectl get daemonsets -A | grep nvidia
 # returns empty
```

- With the following configuration file, I verified that the correct AMI was auto resolved based on the `amiFamily` and the `instanceType`, and that the device plugin daemonset was installed since I included an AL2 nodegroup:

```yaml
---
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: bottlerocket-1-21
  region: us-west-2
  version: '1.21'

nodeGroups:
  - name: br-with-plugin
    instanceType: p3.2xlarge
    desiredCapacity: 1
    amiFamily: Bottlerocket
    iam:
       attachPolicyARNs:
          - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
          - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
          - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
          - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
    ssh:
        allow: true
        publicKeyName: <>
    bottlerocket:
      settings:
        motd: "Hello from eksctl!"
        host-containers:
          admin:
            enabled: true
  - name: al2-with-plugin
    instanceType: p3.2xlarge
    desiredCapacity: 1
    iam:
       attachPolicyARNs:
          - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
          - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
          - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
          - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
    ssh:
        allow: true
        publicKeyName: <>
```

```sh
❯ eksctl get nodegroup --cluster bottlerocket-1-21 -o json | jq '.[].Name, .[].Type, .[].ImageID' -r
al2-with-plugin
br-with-plugin
br-without-plugin
unmanaged
unmanaged
unmanaged
ami-0a2fbc9dc946cfc58
ami-06e0114f5eb33d00c
ami-06e0114f5eb33d00c

~ ❯ kubectl get daemonsets -A | grep nvidia
kube-system   nvidia-device-plugin-daemonset   3         3         3       3            3           <none>          6m30s
```

### Checklist
- [X] Added tests that cover your change (if possible)
- [X] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [X] Manually tested
- [X] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

cPu1 commented on pkg/eks/tasks.go:
```diff
@@ -431,3 +431,23 @@ func (c *ClusterProvider) appendCreateTasksForIAMServiceAccounts(cfg *api.Cluste
+// uses an AMI family which requires the device plugin to be installed as
+// a daemonset
+func clusterSupportsDevicePlugin(cfg *api.ClusterConfig) bool {
+	supported := false
+	for _, ng := range cfg.NodeGroups {
+		if supported {
+			break
+		}
+		supported = ng.NodeGroupBase.GetAMIFamily() != api.NodeImageFamilyBottlerocket
+	}
+	for _, mng := range cfg.ManagedNodeGroups {
+		if supported {
+			break
+		}
+		supported = mng.NodeGroupBase.GetAMIFamily() != api.NodeImageFamilyBottlerocket
+	}
+	return supported
```
Only AL2 GPU AMIs require the device plugin, so the condition can be relaxed. Additionally, `NodeGroupBase` is an embedded field, so `GetAMIFamily` (or `AMIFamily`) can be accessed directly. 

```suggestion
	for _, ng := range cfg.NodeGroups {
		if ng.GetAMIFamily() == api.NodeImageFamilyAmazonLinux2 {
			return true
		}
	}
	for _, mng := range cfg.ManagedNodeGroups {
		if mng.GetAMIFamily() == api.NodeImageFamilyAmazonLinux2 {
			return true
		}
	}
	return false
```

cPu1 commented on pkg/eks/tasks.go:
```diff
@@ -431,3 +431,23 @@ func (c *ClusterProvider) appendCreateTasksForIAMServiceAccounts(cfg *api.Cluste
+// clusterSupportsDevicePlugin returns true if any of the node/managed node groups
+// uses an AMI family which requires the device plugin to be installed as
+// a daemonset
+func clusterSupportsDevicePlugin(cfg *api.ClusterConfig) bool {
```
What do you think about naming this `clusterRequiresDevicePlugin`?

```suggestion
func clusterRequiresDevicePlugin(cfg *api.ClusterConfig) bool {
```

arnaldo2792 commented:
Forced push addresses the feedback left

cPu1 commented:
> Forced push addresses the feedback left

@arnaldo2792, any thoughts on [this comment](https://github.com/weaveworks/eksctl/pull/4809/files#r809728450)?

cPu1 commented on pkg/ami/ssm_resolver.go:
```diff
@@ -108,14 +108,21 @@ func instanceEC2ArchName(instanceType string) string {
 }
 
-func imageType(imageFamily, instanceType string) string {
+func imageType(imageFamily, instanceType string, version string) string {
```
Minor nit:
```suggestion
func imageType(imageFamily, instanceType, version string) string {
```

arnaldo2792 commented:
> > Forced push addresses the feedback left
> 
> @arnaldo2792, any thoughts on [this comment](https://github.com/weaveworks/eksctl/pull/4809/files#r809728450)?

I'm sorry I thought I addressed it in my previous version. I fixed it along with your last nit comment :+1:

aclevername commented on pkg/ami/ssm_resolver_test.go:
```diff
@@ -299,8 +299,56 @@ var _ = Describe("AMI Auto Resolution", func() {
+							resolvedAmi, err = resolver.Resolve(region, version, instanceType, imageFamily)
+						})
+
+						It("does not return an error", func() {
```
```suggestion
						It("errors", func() {
```

arnaldo2792 commented:
Yes @aclevername, AMIs are public, you can pull down the commit and test it yourself :smile: , I'll fix the nit

arnaldo2792 commented:
Forced push addresses the nit pointed out by @aclevername

arnaldo2792 commented:
Forced push is a rebase of `main`

aclevername commented on pkg/eks/tasks.go:
```diff
@@ -356,7 +356,7 @@ func (c *ClusterProvider) ClusterTasksForNodeGroups(cfg *api.ClusterConfig, inst
 		}
 	}
-	if haveNvidiaInstanceType {
+	if haveNvidiaInstanceType && clusterRequiresDevicePlugin(cfg) {
```
Doesn't this mean if my cluster is made up of:
- A regular non-GPU AL2 node
- GPU NVIDIA Bottlerocket node

that it will install the device plugin? Is this the desired behaviour? I just wanna check as I'm not 100% sure I understand when we do/don't need the plugin installed. I would of thought given a non-gpu AL2 only cluster doesn't need it, and a GPU only bottlerocket cluster also doesn't need it, that the combination of the two in the same cluster also wouldn't need it :thinking: 

I validated that it did get installed:
```
cat cluster.yaml
---
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: br-gpu-and-al2-nor
  region: us-west-2
  version: '1.21'

nodeGroups:
  - name: br-nvidia
    instanceType: p3.2xlarge
    desiredCapacity: 1
    amiFamily: Bottlerocket
    bottlerocket:
      settings:
        motd: "Hello from eksctl!"
        host-containers:
          admin:
            enabled: true
  - name: al2-normal
    instanceType: m5.large
    desiredCapacity: 1
```

```
k get daemonsets.apps -A
NAMESPACE     NAME                             DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
kube-system   aws-node                         2         2         2       2            2           <none>          50m
kube-system   kube-proxy                       2         2         2       2            2           <none>          50m
kube-system   nvidia-device-plugin-daemonset   2         2         2       2            2           <none>          39m
```

arnaldo2792 replied:
@aclevername , correct, what you described is what should happen, according to @cPu1 the device plugin should only be installed in AL2 instances, I'll update the code to reflect this. I'd assume that the Neuron device plugin should only be installed in AL2 instances, I'll make the changes to reflect this as well, since I could do a little refactor to remove the function that I added now that I know the conditions required to install the device plugins.

aclevername replied:
> @aclevername , correct, what you described is what should happen, according to @cPu1 the device plugin should only be installed in AL2 instances

I think @cPu1 meant only GPU AL2 Instances? From what I'm reading here https://github.com/weaveworks/eksctl/issues/4760#issuecomment-1040308851

arnaldo2792 replied:
Ok, I'll add a different condition somewhere (if it doesn't exist already), so that folks don't try to install Bottlerocket on inferentia hosts because we don't support it just yet. It is a little weird to me that only AL2 hosts require the NVIDIA GPU device plugin, but it is OK to install the Neuron device plugin on all other hosts, that's why I thought the Neuron device plugin would follow the same rule as the NVIDIA GPU device plugin.

aclevername replied:
I might be wrong, I can't find clear documentation on which nodes actually need to have the plugin running, wether it's just the GPU nodes themselves or all nodes. If it's all nodes then the current code you have is correct. If it's only GPU nodes then we only want to install the nvidia plugin if the cluster contains a AL2 GPU node. 

Does that make senses? I'm sorry I don't have more information :/

arnaldo2792 replied:
Ideally GPU nodes should have the plugin running only if the AMI family supports them, because the device plugin relies on the driver provided by the AMI. Right, now I don't know for sure if AMIs other than AL2 or Bottlerocket support NVIDIA GPUs or Inferentia hardware. The [unit tests](https://github.com/weaveworks/eksctl/blob/57d0afb297742a579391fd3a8fa6ccafcaad3f10/pkg/apis/eksctl.io/v1alpha5/gpu_validation_test.go#L58) suggest that only AL2 supports both NVIDIA and Inferientia instances (and with my change, Bottlerocket will support NVIDIA GPUs).

I think for now my change should only cover installing the device plugin when there is a nodegroup with AL2 with NVIDIA instances, and then we could revisit what other combinations of AMI families/instance types require device plugins.

aclevername reviewed (commented):
thanks for addressing comments @arnaldo2792 . I spotted one thing playing around this morning, please see the above comment :smile: 

we aren't doing a release until Friday so we have some time to work with :smile:

Himangini reviewed (approved):
Nicely done ⭐  LGTM 👍🏻

aclevername commented on pkg/eks/tasks.go:
```diff
@@ -334,29 +334,34 @@ func (c *ClusterProvider) ClusterTasksForNodeGroups(cfg *api.ClusterConfig, inst
+		clusterRequiresNeuronDevicePlugin = clusterRequiresNeuronDevicePlugin ||
+			api.HasInstanceTypeManaged(ng, instanceutils.IsInferentiaInstanceType)
+		// Only AL2 requires the NVIDIA device plugin
+		clusterRequiresNvidiaDevicePlugin = clusterRequiresNvidiaDevicePlugin ||
+			api.HasInstanceTypeManaged(ng, instanceutils.IsNvidiaInstanceType) &&
+				ng.GetAMIFamily() == api.NodeImageFamilyAmazonLinux2
```
my brain gets a bit fuzzy with this chaining, can we add some wrapping brackets?
```suggestion
		clusterRequiresNvidiaDevicePlugin = clusterRequiresNvidiaDevicePlugin || 
			(api.HasInstanceTypeManaged(ng, instanceutils.IsNvidiaInstanceType) &&
				ng.GetAMIFamily() == api.NodeImageFamilyAmazonLinux2)
```

arnaldo2792 commented:
Forced push fixes the nit suggested by @aclevername

arnaldo2792 commented:
Forced push is a rebase of main

aclevername commented:
correcting my earlier mistake, this will be out int `0.86.0-rc0`, not `85` :smile:
